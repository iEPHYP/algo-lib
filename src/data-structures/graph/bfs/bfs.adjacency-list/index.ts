export * from './bfs.adjacency-list';
