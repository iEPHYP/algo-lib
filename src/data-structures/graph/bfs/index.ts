export * from './bfs.adjacency-list';
export * from './bfs.adjacency-matrix';
