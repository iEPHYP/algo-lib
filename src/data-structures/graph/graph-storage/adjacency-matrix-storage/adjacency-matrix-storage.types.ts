export type AdjacencyMatrixGraph = (1 | 0 | undefined)[][];
