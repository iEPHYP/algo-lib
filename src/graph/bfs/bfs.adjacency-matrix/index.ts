export * from './bfs.adjacency-matrix';
