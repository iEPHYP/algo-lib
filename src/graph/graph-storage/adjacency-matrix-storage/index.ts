export * from './adjacency-matrix-storage.types';
