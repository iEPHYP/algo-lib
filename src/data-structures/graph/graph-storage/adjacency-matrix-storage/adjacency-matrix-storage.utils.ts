import { getNormalizedEdges, RawEdges } from '../edges-storage';
import { AdjacencyMatrixGraph } from './adjacency-matrix-storage.types';

export const convertEdgesToAdjacencyMatrixGraph = (
  rawEdges: RawEdges,
  // TODO: test directed graph
  indirected = true
) => {
  const edges = getNormalizedEdges(rawEdges);

  const graph: AdjacencyMatrixGraph = [];

  edges.forEach(({ vertexOutIndex, vertexInIndex }) => {
    graph[vertexOutIndex] = graph[vertexOutIndex] || [];
    const vertexOutRow = graph[vertexOutIndex];

    vertexOutRow[vertexInIndex] = 1;

    if (indirected) {
      graph[vertexInIndex] = graph[vertexInIndex] || [];
      const vertexInRow = graph[vertexInIndex];

      vertexInRow[vertexOutIndex] = 1;
    }
  });

  return graph;
};
