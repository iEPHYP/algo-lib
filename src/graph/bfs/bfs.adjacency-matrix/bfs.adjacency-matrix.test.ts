import { convertEdgesToAdjacencyMatrixGraph } from '../../graph-storage/adjacency-matrix-storage/adjacency-matrix-storage.utils';
import { indirectConnectedGraphRawEdges } from '../../rawEdges.mock';
import { adjacencyMatrixGraphBfs } from './bfs.adjacency-matrix';

test('adjacencyMatrixGraphDfs should print visited vertices in the right order', () => {
  const graph = convertEdgesToAdjacencyMatrixGraph(indirectConnectedGraphRawEdges);

  const visitedVertexIndices: number[] = [];

  adjacencyMatrixGraphBfs(graph, (vertexIndex) => {
    visitedVertexIndices.push(vertexIndex);
  });

  expect(visitedVertexIndices).toMatchSnapshot();
});
