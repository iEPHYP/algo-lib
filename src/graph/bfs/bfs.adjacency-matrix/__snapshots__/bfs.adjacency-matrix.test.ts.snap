// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adjacencyMatrixGraphDfs should print visited vertices in the right order 1`] = `
[
  0,
  1,
  3,
  5,
  6,
  4,
  2,
  7,
]
`;

exports[`adjacencyMatrixGraphDfs should print visited vertices in the right order 2`] = `
[
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
]
`;
